### ISSUE TITLE
Unable to connect to SOAP port after updating to TLS 1.1+

### ISSUE THREAD

**Issue by JJinMaine** [yannlambret/websphere-nagios#1]
Hi Yann,

Last night I tested out some changes to WebSphere 7.0.0.41 to accept TLS 1.1 and 1.2 and to not accept SSLv3 or TLS 1.0. Ever since I restarted those WAS processes, the wasagent has been failing. I used [these instructions here](https://developer.ibm.com/answers/questions/206952/how-do-i-configure-websphere-application-server-ss.html) to do the changes. Any ideas as to why this is happening or how to resolve? Thanks!

\- Jim

```
2017-01-25 12:21:45.341:INFO:oejs.Server:jetty-7.6.2.v20120308
2017-01-25 12:21:45.988:INFO:oejsh.ContextHandler:started o.e.j.s.ServletContextHandler{/wasagent,null}
2017-01-25 12:21:46.011:INFO:oejs.AbstractConnector:Started SocketConnector@127.0.0.1:10500
Jan 25, 2017 12:21:54 PM null null
WARNING: ADMC0046W: The system failed to load properties from com.ibm.SOAP.configURL: file:null/properties/soap.client.props
Jan 25, 2017 12:21:59 PM null null
WARNING: Could not find tmx4jTransform.jar in null/etc/tmx4jTransform.jar - Interoperability to older versions of WebSphere is disabled
Jan 25, 2017 12:22:04 PM null null
INFO: CWPKI0027I: Disabling default hostname verification for HTTPS URL connections.
Jan 25, 2017 12:22:04 PM null null
INFO: Client code attempting to load security configuration
com.ibm.websphere.management.exception.ConnectorException: ADMC0016E: The system cannot create a SOAP connector to connect to host localhost at port 8880.
        at com.ibm.websphere.management.AdminClientFactory.createAdminClientPrivileged(AdminClientFactory.java:636)
        at com.ibm.websphere.management.AdminClientFactory.access$000(AdminClientFactory.java:128)
        at com.ibm.websphere.management.AdminClientFactory$1.run(AdminClientFactory.java:211)
        at com.ibm.ws.security.util.AccessController.doPrivileged(AccessController.java:63)
        at com.ibm.websphere.management.AdminClientFactory.createAdminClient(AdminClientFactory.java:207)
        at net.wait4it.nagios.wasagent.core.WASClientProxy.init(Unknown Source)
        at net.wait4it.nagios.wasagent.core.WASStatsManager.process(Unknown Source)
        at net.wait4it.nagios.wasagent.core.WASServlet.doGet(Unknown Source)
        at net.wait4it.nagios.wasagent.core.WASServlet.doPost(Unknown Source)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:738)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:831)
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:565)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:479)
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:225)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1031)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:406)
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:186)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:965)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:117)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:111)
        at org.eclipse.jetty.server.Server.handle(Server.java:349)
        at org.eclipse.jetty.server.AbstractHttpConnection.handleRequest(AbstractHttpConnection.java:452)
        at org.eclipse.jetty.server.BlockingHttpConnection.handleRequest(BlockingHttpConnection.java:47)
        at org.eclipse.jetty.server.AbstractHttpConnection.content(AbstractHttpConnection.java:894)
        at org.eclipse.jetty.server.AbstractHttpConnection$RequestHandler.content(AbstractHttpConnection.java:948)
        at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:857)
        at org.eclipse.jetty.http.HttpParser.parseAvailable(HttpParser.java:235)
        at org.eclipse.jetty.server.BlockingHttpConnection.handle(BlockingHttpConnection.java:66)
        at org.eclipse.jetty.server.bio.SocketConnector$ConnectorEndPoint.run(SocketConnector.java:254)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:599)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:534)
        at java.lang.Thread.run(Thread.java:767)
Caused by: java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:56)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:39)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:527)
        at com.ibm.websphere.management.AdminClientFactory.createAdminClientPrivileged(AdminClientFactory.java:458)
        ... 31 more
Caused by: com.ibm.websphere.management.exception.ConnectorNotAvailableException: [SOAPException: faultCode=SOAP-ENV:Client; msg=Error parsing HTTP status line &quot;&quot;: java.util.NoSuchElementException; targetException=java.lang.IllegalArgumentException: Error parsing HTTP status line "": java.util.NoSuchElementException]
        at com.ibm.ws.management.connector.soap.SOAPConnectorClient.reconnect(SOAPConnectorClient.java:415)
        at com.ibm.ws.management.connector.soap.SOAPConnectorClient.<init>(SOAPConnectorClient.java:227)
        ... 36 more
Caused by: [SOAPException: faultCode=SOAP-ENV:Client; msg=Error parsing HTTP status line &quot;&quot;: java.util.NoSuchElementException; targetException=java.lang.IllegalArgumentException: Error parsing HTTP status line "": java.util.NoSuchElementException]
        at org.apache.soap.transport.http.SOAPHTTPConnection.send(SOAPHTTPConnection.java:475)
        at org.apache.soap.rpc.Call.WASinvoke(Call.java:510)
        at com.ibm.ws.management.connector.soap.SOAPConnectorClient$4.run(SOAPConnectorClient.java:379)
        at com.ibm.ws.security.util.AccessController.doPrivileged(AccessController.java:118)
        at com.ibm.ws.management.connector.soap.SOAPConnectorClient.reconnect(SOAPConnectorClient.java:372)
        ... 37 more
Jan 25, 2017 12:22:36 PM null null

```

**Comment by JJinMaine** [comment]
Ok, I think I found a solution for this and there are two parts:

1. These two options need to be added to the JAVA_OPTS in your run.sh:

`-Dcom.ibm.jsse2.overrideDefaultProtocol=SSL_TLSv2 -Djava.security.properties=java.security`

From the trace I ran, it appears that SSLv3 is the default call protocol being made so you need to override it with something more secure like the TLS 1.X protocols if you've disabled SSLv3 in your Default SSL Settings in the WebSphere cell you're connecting to. In my java.security we have this entry in place:

`jdk.tls.disabledAlgorithms=TLSv1,SSLv3`

So I know that even though I'm overriding with SSL_TLSv2, only TLSv1.1 and 1.2 will actually be usable - TLSv1.0 won't be.

2. Copy your java.security from your $JAVA_HOME/jre/lib/security/java.security location to your wasagent folder so that wasagent.properties and java.security are in the same location. The reason is because we're going to modify the Default SSL Socket Factories and if your wasagent is on the same server as your WebSphere installation, you may not want to change that java.security file. If the JAVA_HOME you're using belongs to wasagent and there's no risk of affecting a WebSphere installation then maybe nevermind. However if you're on the same server it's probably better if you copy it to the wasagent folder and override settings on the command line. If you've been disabling SSLv3 and TLSv1.0 like I have, you've already been in java.security lately so it won't be foreign to you. Go in java.security and comment out the WebSphere socket factories and uncomment Default JSSE socket factories so it looks like this:

```
# Default JSSE socket factories
ssl.SocketFactory.provider=com.ibm.jsse2.SSLSocketFactoryImpl
ssl.ServerSocketFactory.provider=com.ibm.jsse2.SSLServerSocketFactoryImpl
# WebSphere socket factories (in cryptosf.jar)
#ssl.SocketFactory.provider=com.ibm.websphere.ssl.protocol.SSLSocketFactory
#ssl.ServerSocketFactory.provider=com.ibm.websphere.ssl.protocol.SSLServerSocketFactory
```

You may want to do your own testing with TLS 1.1 and TLS 1.2 but this worked for me. Sorry if my explanation was a run-on or too wordy. Let me know if you have any questions - thanks!

**Comment by yannlambret** [comment]
Hi Jim,

Sorry for the late answer. I haven't be using WebSphere for a long time now, and I'm not able to test or validate anything myself. Feel free to open a pull request and I'll be glad to merge your fix.

Yann
